sink-mongo/src/lib.rs

```diff
@@ -0,0 +1,88 @@
+use apibara_core::node::v1alpha2::{Cursor, DataFinality};
+use apibara_sink_common::Sink;
+use async_trait::async_trait;
+use mongodb::bson::{doc, to_bson, Document};
+
+use mongodb::{error::Error, options::ClientOptions, Client, Collection};
+
+use serde_json::{Value};
+use tracing::{info, warn};
+
+pub struct MongoSink {
+    collection: Collection<Document>,
+}
+
+impl MongoSink {
+    pub async fn new(
+        mongo_url: String,
+        db_name: String,
+        collection_name: String,
+    ) -> Result<Self, Error> {
+        // Parse a connection string into an options struct.
+        let client_options = ClientOptions::parse(mongo_url).await?;
+
+        let client = Client::with_options(client_options)?;
+        let db = client.database(&db_name);
+        let collection = db.collection::<Document>(&collection_name);
+
+        Ok(Self { collection })
+    }
+}
+
+#[async_trait]
+impl Sink for MongoSink {
+    type Error = Error;
+
+    async fn handle_data(
+        &mut self,
+        cursor: &Option<Cursor>,
+        end_cursor: &Cursor,
+        finality: &DataFinality,
+        batch: &Value,
+    ) -> Result<(), Self::Error> {
+        info!(
+            cursor = ?cursor,
+            end_cursor = ?end_cursor,
+            finality = ?finality,
+            batch = ?batch,
+            "inserting data to mongo"
+        );
+
+        // convert to i64 because that's the maximum bson can handle
+        let block_number = i64::try_from(end_cursor.order_key).unwrap_or(-1);
+
+        match batch {
+            Value::Array(array) => {
+                let documents: Vec<Document> = array
+                    .iter()
+                    .map(
+                        // Use unwrap because if the data couldn't be converted to bson
+                        // we can't write it to Mongo and thus there is point to continue
+                        |element| doc! {"data": to_bson(element).unwrap(), "_cursor": block_number},
+                    )
+                    .collect();
+                self.collection.insert_many(documents, None).await?;
+            }
+            value => {
+                warn!(value = ?value, "batch value is not an array");
+                // Use unwrap because if the data couldn't be converted to bson
+                // we can't write it to Mongo and thus there is point to continue
+                let document = doc! {"data": to_bson(value).unwrap(), "_cursor": block_number};
+                self.collection.insert_one(document, None).await?;
+            }
+        }
+        Ok(())
+    }
+
+    async fn handle_invalidate(&mut self, cursor: &Option<Cursor>) -> Result<(), Self::Error> {
+        info!(cursor = ?cursor, "invalidating data from mongo");
+
+        if let Some(cursor) = cursor {
```

Comment:
If the cursor is `None`, then the chain reorged including the genesis block so we should delete all items in the collection.